### ISSUE TITLE
Effects of setColor() on ByteProcessor vs. ColorProcessor

### ISSUE THREAD

**Issue by imagingbook** [imagej/imagej1#36]
It appears that the ``setColor()`` method does not consider RGB component weights properly when setting the associated (i.e., gray) draw value for ``ByteProcessor``. Drawing operations should yield the same brightness values when (a) drawing to a grayscale image or (b) drawing to a color image and subsequently converting to a grayscale image. However, results are vastly different.

The following plugin demonstrates this behavior: the values in ``bp`` should be the same as in the converted color image ``cpc``. Test images are attached.

````
import java.awt.Color;

import ij.ImagePlus;
import ij.plugin.PlugIn;
import ij.process.ByteProcessor;
import ij.process.ColorProcessor;
import ij.process.ImageProcessor;

public class Color_Draw_Test implements PlugIn {
	static int width = 200;
	static int height = 200;
	
	static Color circleColor = Color.blue;
	static Color backgroundColor = Color.cyan;

	@Override
	public void run(String arg0) {

		ImageProcessor cp = new ColorProcessor(width, height);
		ImageProcessor bp = new ByteProcessor(width, height);
		cp.setColor(backgroundColor); cp.fill();
		bp.setColor(backgroundColor); bp.fill();
		
		cp.setColor(circleColor);
		cp.fillOval(25, 50, 150, 100);
		
		bp.setColor(circleColor);
		bp.fillOval(25, 50, 150, 100);
		
		ImageProcessor cpc = cp.convertToByteProcessor();

		new ImagePlus("cp", cp).show();
		new ImagePlus("cpc", cpc).show();
		new ImagePlus("bp", bp).show();
	}
}
````

If this is agreed to need fixing I'd be happy to look into it.
--Wilhelm

cp ![cp](https://cloud.githubusercontent.com/assets/3172121/26033384/7768af08-38ab-11e7-9030-c7de3c05caca.png)
cpc ![cpc](https://cloud.githubusercontent.com/assets/3172121/26033385/7a82d29a-38ab-11e7-8f63-db4a403c58bf.png)
bp ![bp](https://cloud.githubusercontent.com/assets/3172121/26033383/73d9b13e-38ab-11e7-9ced-ed29edba5411.png)

**Comment by rasband** [comment]
This bug should be fixed in the latest ImageJ daily build (1.51o11). This fix adds this code to the ImageProcessor.getBestIndex() method:

```
        if (r1!=g1 || g1!=b1 || r1!=b1) {
            double[] w = ColorProcessor.getWeightingFactors();
            double oneThird = 1d/3d;
            r1 = (int)Math.round(r1*(w[0]/oneThird));
            g1 = (int)Math.round(g1*(w[1]/oneThird));
            b1 = (int)Math.round(b1*(w[2]/oneThird));
        }       

```

**Comment by imagingbook** [comment]
Thanks. I do not fully understand the details of the above code but isn't ``w[0]/oneThird`` the same as ``w[0]*3``?

**Comment by rasband** [comment]
You are correct. This is the fix that is in the latest daily build (1.51o13).

```
        if (!(r1==g1&&g1==b1&&r1==b1) && !isColorLut()) {
            double[] w = ColorProcessor.getWeightingFactors();
            r1 = (int)Math.round(3*r1*w[0]);
            g1 = (int)Math.round(3*g1*w[1]);
            b1 = (int)Math.round(3*b1*w[2]);
        }       

```
